driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java

```diff
@@ -63,6 +63,9 @@ void refresh(Cluster cluster,
                 assert rows.keyspaces != null;
                 Map<String, KeyspaceMetadata> keyspaces = buildKeyspaces(rows, cassandraVersion, cluster);
                 updateKeyspaces(metadata, metadata.keyspaces, keyspaces, targetKeyspace);
+                // If we rebuild all from scratch or have an updated keyspace, rebuild the token map
+                // since some replication on some keyspace may have changed
+                metadata.rebuildTokenMap();
```

Comment:
I am still a bit worried bout this call being synchronous, but not atomic. Here, right before the call to `rebuildTokenMap`, the field `Metadata.keyspaces` would contain the changes, but not the field `Metadata.tokenMap`. There is still a slight, albeit improbable, chance that an observer could retrieve inconsistent data. What makes it difficult to fix is that updates to `Metadata.keyspaces` are made on-the-go.